src/main/java/bio/terra/service/common/gcs/GcsUriUtils.java

```diff
@@ -0,0 +1,88 @@
+package bio.terra.service.common.gcs;
+
+import bio.terra.service.dataset.exception.InvalidUriException;
+import org.apache.commons.lang3.StringUtils;
+
+public final class GcsUriUtils {
+  public static class GsUrlParts {
+    private String bucket;
+    private String path;
+    private boolean isWildcard = false;
+
+    public String getBucket() {
+      return bucket;
+    }
+
+    public GsUrlParts bucket(String bucket) {
+      this.bucket = bucket;
+      return this;
+    }
+
+    public String getPath() {
+      return path;
+    }
+
+    public GsUrlParts path(String path) {
+      this.path = path;
+      return this;
+    }
+
+    public boolean getIsWildcard() {
+      return isWildcard;
+    }
+
+    public GsUrlParts isWildcard(boolean isWildcard) {
+      this.isWildcard = isWildcard;
+      return this;
+    }
+  }
+
+  /**
+   * Parse a Google Cloud Storage URI into its component pieces
+   *
+   * @param uri of type gs://<bucket_name>/<file_path_inside_bucket>
+   * @return Object representing uri pieces
+   */
+  public static GsUrlParts parseBlobUri(String uri) {
+    String protocol = "gs://";
+    if (!StringUtils.startsWith(uri, protocol)) {
```

Comment:
I would remove `StringUtils` from all of this code; the main benefit is that it supports `null` and case insensitive comparisons, and neither case is useful here. If we do expect `null` we can test for that first here and assume it afterwards.
```suggestion
    if (uri == null || !uri.startsWith(protocol)) {
```

Reply:
Stringutils has a convenient substringBefore method, if I got rid of that I'd essentially have to copy paste the function logic over. I think we should keep it

```
    public static String substringBefore(final String str, final String separator) {
        if (isEmpty(str) || separator == null) {
            return str;
        }
        if (separator.isEmpty()) {
            return EMPTY;
        }
        final int pos = str.indexOf(separator);
        if (pos == INDEX_NOT_FOUND) {
            return str;
        }
        return str.substring(0, pos);
    }
```

Reply:
I don't think any of the logic there is needed:
- the uri string is not empty
- the separator is not empty
- there must be at least one `/` in the uri or it wouldn't be a valid gs path

```
int slashPos = str.indexOf('/');
if (slashPos == -1) { throw ... }
String bucket = str.subString(0, slashPos);
String path = str.subString(slashPos + 1);
```